### ISSUE TITLE
Sokoban model response turned '!!!!!!!' after steps for PPO training

### ISSUE THREAD

**Issue by scris** [ZihanWang314/RAGEN#52]
Thanks for the brilliant and inspiring work, but I met some problem during training:

The response of sokoban PPO training is correct at first:

<img width="588" alt="Image" src="https://github.com/user-attachments/assets/92c52371-0659-4ceb-a6ce-53c01b141304" />

But after some steps, it went weird as '!!!!!', stay score -5.5 and did not recover from then on:

<img width="598" alt="Image" src="https://github.com/user-attachments/assets/c599d20b-be26-4041-a593-70972aa33416" />

For this script (sokoban base prompt in `cmd.md`), the result turned to be '!!!!!' at epoch 0 step 63.

``` bash
bash train.sh sokoban \
    model.base_model=Qwen/Qwen2.5-0.5B-Instruct \
    training.train_batch_size=4 \
    training.max_turns=5 \
    training.n_rollout=8 \
    training.micro_batch_size=4 \
    training.ppo_batch_size=32 \
    optimization.kl_coef=0.001 \
    optimization.adv_estimator=brpo
```

For this script, the result turned to be '!!!!!' at epoch 0 step 64.

``` bash
bash train.sh sokoban \
    model.base_model=Qwen/Qwen2.5-0.5B-Instruct
```

For this script, the result turned to be '!!!!!' at epoch 0 step 68.

``` bash
bash train.sh sokoban \
    model.base_model=Qwen/Qwen2.5-0.5B-Instruct \
    training.train_batch_size=4 \
    training.max_turns=5 \
    training.n_rollout=4 \
    training.micro_batch_size=2
```

For this script, the result turned to be '!!!!!' at epoch 0 step 95 for 0.5B, and epoch 0 step 51 for 1.5B.

``` bash
bash train.sh sokoban \
    model.base_model=Qwen/Qwen2.5-[0.5B|1.5B]-Instruct \
    training.train_batch_size=4 \
    training.max_turns=5 \
    training.n_rollout=8
```

Also, critic/reward/mean fails to go higher after approximately step 30 in my tests.

These tests are done with Python 3.10, PyTorch 2.4, CUDA 12.1 on one Nvidia A100 or one Nvidia A6000. Any suggestions on how to solve this problem? Thanks very much!

**Comment by ZihanWang314** [comment]
This is what we are dealing with in our coding report. We believe this is a problem with the algorithm rather than implementation -> when the model is trained enough steps, there will be some "model collapse" issues.
Please stay tuned for our updates!

**Comment by scris** [comment]
Thanks for your swift reply! I'm really interested in your project.

**Comment by scris** [comment]
These experiments above are for replicating https://github.com/ZihanWang314/RAGEN/blob/main/public/loss_curve.png. But clearly I failed to do that with an almost-up-to-date code (db0fa39 on 2.21), so I tried another version (f991d1d on 1.28, the version when `loss_curve.png` is uploaded), and with Qwen 2.5 0.5B Instruct, I got these results:

<img width="923" alt="Image" src="https://github.com/user-attachments/assets/7b607d33-bc7e-4354-9ecb-e85bccea410d" />

I found 2.21 code resulted in -5.5 straight line (as response turned '!!!!!!', exp3 and exp16), but 1.28 code can successfully run for more steps (exp20). I also noticed that the 1.28 code use max_step = 5 (easier data) and penalize invalid response with -0.1, which are different from 2.21. I redo the experiment with max_step = 10 and -1 penalty (exp22), which are identical to 2.21's setting, and no mode collapse is found either, and model seems to be learning more comparing to exp3 and exp16 in the perspective of the content quality of the trajectory model response, although the training process is a bit slower.

May I ask what's the difference of the two code versions, 2.21 db0fa39 and 1.28 f991d1d? Thank you very much.

**Comment by ZihanWang314** [comment]
Thanks for the update, sorry for the late reply! I think the main difference lies in the default config. Over the past days we are trying to implement as good as possible baseline config so everyone could be able to serve them with less effort into their own tasks. However there do sometimes will be tradeoffs. We find the current observations for now:

1. GAE is needed
2. need larger PPO batch size
3. need 1.5B level models

I hope it helps!

**Comment by ZihanWang314** [comment]
We are not very sure now, haha.
But we find other developers do share similar observations, like the Open-Reasoner-Zero team. Feel free to check https://x.com/rosstaylor90/status/1892664646890312125

**Comment by scris** [comment]
Thanks! I'll check that.

**Comment by ZihanWang314** [comment]
https://github.com/ZihanWang314/RAGEN/issues/58
Please refer to this issue. This is actually fixed by changing to the latest implementation of veRL. Thanks!

**Comment by scris** [comment]
However, I still witnessed some problem on my side for GRPO running with 1.5B and 0.5B models (The result IS better, but issue is NOT solved), do you have any idea? Thanks!

<img width="2019" alt="Image" src="https://github.com/user-attachments/assets/f4862e71-daa7-4809-8f30-8772c5023725" />

``` bash
bash train.sh sokoban \
    model.base_model=Qwen/Qwen2.5-0.5B-Instruct \
    training.train_batch_size=4 \
    training.max_turns=5 \
    training.n_rollout=4 \
    training.micro_batch_size=2
```

**Comment by ZihanWang314** [comment]
We tested on bandit tasks but not on Sokoban (will also push a mimor bug for bandit tasks). Let me reopen this issue and see if there's any other problems.

**Comment by ZihanWang314** [comment]
By the way, it seems that PPO does not have issues at your side. Could you please try using PPO?
We noticed someone in Chinese QA platforms that GRPO could be a biased estimator. We are not sure if it is because of this kind of problem. Please feel free to check this https://zhuanlan.zhihu.com/p/28735759256

**Comment by scris** [comment]
Thanks! I'm just using some time every day trying to figure out GRPO :) 
And thanks for your project :)

**Comment by carrot-codeing** [comment]
The same, I have a question: Is it necessary to train the model using the complete 10k dataset provided by the author? How can I determine when to stop the training? Also, I would like to know how to evaluate the performance of the trained model. How can I obtain the visualizations mentioned above?

**Comment by MHlk** [comment]
Same problem! Actually I used the **GAE** instead of GRPO, but this problem still appears at epoch 0 step 100 when I was training qwen2.5-1.5b-instruct.

This is my config setting:
```
### Training parameters
training:
  train_data_num: null
  val_data_num: 50
  micro_batch_size: 16
  train_batch_size: 32
  val_batch_size: 10
  ppo_batch_size: 128
  max_start_length: 400
  max_response_length: 400
  max_obs_length: 200
  max_turns: 5
  rollout_tp_size: 4
  n_rollout: 16
  total_epochs: 5
  temperature: 0.7
  use_sft: false
  use_kl_loss: true # this means actor KL Loss
  no_think_rl: false
  state_masking: false
  binary_reward: false
  mask_state: false
  length_penalty: false
  ref_update_steps: null # every ref_update_steps, the reference model will be updated using the latest actor model
  total_training_steps: null

### Optimization parameters
optimization:
  actor_lr: 1e-6
  critic_lr: 1e-5
  kl_coef: 0.001
  kl_loss_type: low_var_kl
  adv_estimator: gae
  gpu_memory_utilization: 0.4
```

**Comment by scris** [comment]
@MHlk Ouch, I think this problem will only happen in non-PPO methods like GRPO or REINFORCE++.

**Comment by MHlk** [comment]
> [@MHlk](https://github.com/MHlk) Ouch, I think this problem will only happen in non-PPO methods like GRPO or REINFORCE++.

I'm not sure if it because my training parameters aren't suitable or the model scale is too small🤣 I may change a larger model and different training parameters to see if this phenomenon still occurs.

**Comment by ZihanWang314** [comment]
Hi, just to let you know that we've updated a new version of code with improved stability training stability, and we wrote a [paper](https://github.com/RAGEN-AI/RAGEN/blob/main/RAGEN.pdf) to discuss this issue. The model collapse is largely postponed or diminished in the latest setting on 0.5B Qwen instruct models. Thanks!

**Comment by MHlk** [comment]
> Hi, just to let you know that we've updated a new version of code with improved stability training stability, and we wrote a [paper](https://github.com/RAGEN-AI/RAGEN/blob/main/RAGEN.pdf) to discuss this issue. The model collapse is largely postponed or diminished in the latest setting on 0.5B Qwen instruct models. Thanks!

Thanks for your timely reply! I'll try the new code :)
